# Análisis Exploratorio de Datos - Wall-Following Robot Navigation

**Estudiante:** Cristian Zurita

## Objetivo

Realizar la carga y exploración inicial del dataset Wall-Following Robot Navigation,
identificando su estructura, tipos de datos, variable objetivo y distribución de clases.

In [1]:
from pathlib import Path
import pandas as pd

In [2]:
project_root = Path.cwd()

if project_root.name == "notebooks":
    project_root = project_root.parent

data_path = project_root / "data" / "raw" / "sensor_readings_24.data"

print("Raíz del proyecto:", project_root)
print("Ruta del dataset:", data_path)
print("¿Existe el dataset?:", data_path.exists())

Raíz del proyecto: /mnt/DiscoD/U/MachineLearning/ml-practica1-zurita
Ruta del dataset: /mnt/DiscoD/U/MachineLearning/ml-practica1-zurita/data/raw/sensor_readings_24.data
¿Existe el dataset?: True


In [3]:
sensor_columns = [f"US{i}" for i in range(1, 25)]
columns = sensor_columns + ["Class"]

print(columns)
print("Número de columnas:", len(columns))

['US1', 'US2', 'US3', 'US4', 'US5', 'US6', 'US7', 'US8', 'US9', 'US10', 'US11', 'US12', 'US13', 'US14', 'US15', 'US16', 'US17', 'US18', 'US19', 'US20', 'US21', 'US22', 'US23', 'US24', 'Class']
Número de columnas: 25


In [4]:
df = pd.read_csv(
    data_path,
    header=None,
    names=columns
)

df.head()

,US1,US2,US3,US4,US5,US6,US7,US8,US9,US10,...,US16,US17,US18,US19,US20,US21,US22,US23,US24,Class
0,0.438,0.498,3.625,3.645,5.0,2.918,5.0,2.351,2.332,2.643,...,0.593,0.502,0.493,0.504,0.445,0.431,0.444,0.440,0.429,Slight-Right-Turn
1,0.438,0.498,3.625,3.648,5.0,2.918,5.0,2.637,2.332,2.649,...,0.592,0.502,0.493,0.504,0.449,0.431,0.444,0.443,0.429,Slight-Right-Turn
2,0.438,0.498,3.625,3.629,5.0,2.918,5.0,2.637,2.334,2.643,...,0.593,0.502,0.493,0.504,0.449,0.431,0.444,0.446,0.429,Slight-Right-Turn
3,0.437,0.501,3.625,3.626,5.0,2.918,5.0,2.353,2.334,2.642,...,0.593,0.502,0.493,0.504,0.449,0.431,0.444,0.444,0.429,Slight-Right-Turn
4,0.438,0.498,3.626,3.629,5.0,2.918,5.0,2.640,2.334,2.639,...,0.592,0.502,0.493,0.504,0.449,0.431,0.444,0.441,0.429,Slight-Right-Turn


## Dimensiones del dataset

En esta sección se verifica el tamaño del conjunto de datos, es decir,
cuántas observaciones y cuántas variables contiene.

In [5]:
print("Dimensiones del dataset:", df.shape)
print("Número de filas:", df.shape[0])
print("Número de columnas:", df.shape[1])

Dimensiones del dataset: (5456, 25)
Número de filas: 5456
Número de columnas: 25


## Tipos de datos

Se revisa el tipo de dato de cada columna para distinguir las variables numéricas
de la variable objetivo categórica.

In [6]:
df.dtypes

US1      float64
US2      float64
US3      float64
US4      float64
US5      float64
US6      float64
US7      float64
US8      float64
US9      float64
US10     float64
US11     float64
US12     float64
US13     float64
US14     float64
US15     float64
US16     float64
US17     float64
US18     float64
US19     float64
US20     float64
US21     float64
US22     float64
US23     float64
US24     float64
Class        str
dtype: object

## Estadísticas descriptivas

Se calculan estadísticas básicas de las variables numéricas para conocer
su distribución, valores mínimos, máximos, media y dispersión.

In [7]:
df.describe()

,US1,US2,US3,US4,US5,US6,US7,US8,US9,US10,...,US15,US16,US17,US18,US19,US20,US21,US22,US23,US24
count,5456.000000,5456.000000,5456.000000,5456.000000,5456.000000,5456.000000,5456.000000,5456.000000,5456.000000,5456.000000,...,5456.000000,5456.000000,5456.000000,5456.000000,5456.00000,5456.000000,5456.000000,5456.000000,5456.000000,5456.000000
mean,1.471617,2.327043,2.489347,2.796501,2.958552,2.893073,3.351113,2.540397,3.125621,2.832386,...,2.205772,1.202111,0.989831,0.910273,1.05811,1.076320,1.015923,1.778034,1.555045,1.578508
std,0.802801,1.410146,1.247435,1.309368,1.339225,1.282575,1.413692,1.111554,1.356965,1.307843,...,1.715435,1.098568,0.942075,0.889527,1.14463,1.141498,0.887439,1.571686,1.291447,1.150480
min,0.400000,0.437000,0.470000,0.833000,1.120000,1.114000,1.122000,0.859000,0.836000,0.810000,...,0.495000,0.424000,0.373000,0.354000,0.34000,0.355000,0.380000,0.370000,0.367000,0.377000
25%,0.921000,1.362000,1.538750,1.731000,1.774000,1.785750,1.930750,1.618000,1.799750,1.636000,...,0.860000,0.690000,0.581000,0.529750,0.52300,0.541750,0.567000,0.743000,0.792000,0.884000
50%,1.335000,1.904500,2.064000,2.458000,2.667000,2.682500,3.225500,2.172000,2.802000,2.679000,...,1.328500,0.803000,0.738000,0.685000,0.69100,0.693000,0.764000,1.030500,1.071000,1.289000
75%,1.814000,2.681500,2.739250,4.093500,4.314500,3.835250,5.000000,3.193000,5.000000,3.526250,...,4.436250,1.159000,0.913000,0.837000,0.85700,0.863000,1.002250,2.068250,1.559500,1.657250
max,5.000000,5.025000,5.029000,5.017000,5.000000,5.005000,5.008000,5.087000,5.000000,5.022000,...,5.000000,5.000000,5.000000,5.000000,5.00000,5.000000,5.000000,5.000000,5.000000,5.000000


## Distribución de la variable objetivo

Se analiza cuántas observaciones pertenecen a cada clase de navegación
para identificar si el conjunto de datos está balanceado o desbalanceado.

In [8]:
df["Class"].value_counts()

Class
Move-Forward         2205
Sharp-Right-Turn     2097
Slight-Right-Turn     826
Slight-Left-Turn      328
Name: count, dtype: int64

## Valores faltantes

Se verifica si existen datos ausentes en alguna de las variables.
Los valores faltantes deben identificarse antes de entrenar modelos,
ya que pueden requerir eliminación o imputación.

In [9]:
missing_values = df.isnull().sum()

print("Valores faltantes por columna:")
print(missing_values)

print("\nTotal de valores faltantes:", missing_values.sum())

Valores faltantes por columna:
US1      0
US2      0
US3      0
US4      0
US5      0
US6      0
US7      0
US8      0
US9      0
US10     0
US11     0
US12     0
US13     0
US14     0
US15     0
US16     0
US17     0
US18     0
US19     0
US20     0
US21     0
US22     0
US23     0
US24     0
Class    0
dtype: int64

Total de valores faltantes: 0


## Registros duplicados

Se comprueba la existencia de filas completamente repetidas.
Los duplicados pueden influir en el análisis y deben identificarse
antes de decidir si es necesario eliminarlos.

In [10]:
duplicate_rows = df.duplicated().sum()

print("Número de filas duplicadas:", duplicate_rows)

Número de filas duplicadas: 0


## Balance porcentual de clases

Además del número de observaciones por clase, se calcula su proporción
porcentual para evaluar con mayor claridad el nivel de desbalance
de la variable objetivo.

In [11]:
class_counts = df["Class"].value_counts()
class_percentages = df["Class"].value_counts(normalize=True) * 100

class_balance = pd.DataFrame({
    "Cantidad": class_counts,
    "Porcentaje (%)": class_percentages.round(2)
})

class_balance

,Cantidad,Porcentaje (%)
Class,,
Move-Forward,2205,40.41
Sharp-Right-Turn,2097,38.43
Slight-Right-Turn,826,15.14
Slight-Left-Turn,328,6.01


### Interpretación del balance de clases

La distribución de la variable objetivo no es uniforme. Las clases `Move-Forward`
(40.41 %) y `Sharp-Right-Turn` (38.43 %) concentran la mayor parte de las
observaciones, mientras que `Slight-Right-Turn` representa el 15.14 % y
`Slight-Left-Turn` únicamente el 6.01 %.

Por lo tanto, el dataset presenta desbalance de clases. Este aspecto deberá
considerarse posteriormente durante la división de los datos y la evaluación
de los modelos, especialmente para evitar que un buen rendimiento global
oculte un desempeño deficiente en las clases menos representadas.

## Conclusiones de la exploración inicial

El dataset contiene **5.456 observaciones** y **25 columnas**, de las cuales
24 corresponden a lecturas de sensores ultrasónicos (`US1` a `US24`) y una
corresponde a la variable objetivo `Class`.

Las variables de los sensores son numéricas y la variable objetivo es
categórica. No se encontraron valores faltantes ni filas duplicadas.

La variable objetivo presenta cuatro clases de navegación:

- `Move-Forward`: 2205 observaciones (40.41 %).
- `Sharp-Right-Turn`: 2097 observaciones (38.43 %).
- `Slight-Right-Turn`: 826 observaciones (15.14 %).
- `Slight-Left-Turn`: 328 observaciones (6.01 %).

La distribución de clases no es uniforme, por lo que existe desbalance,
especialmente en la clase `Slight-Left-Turn`. Este aspecto deberá considerarse
posteriormente durante la división de los datos y la evaluación de los modelos.